### ISSUE TITLE
Why don't dynamically sized matrices implement `Hash`?

### ISSUE THREAD

**Issue by mwillsey** [rustsim/nalgebra#508]
This is preventing me from putting dynamically sized matrices in a hashmap. Any reason why this is the case? 

Thanks!

Minimal example:

```rust
fn foo(mat: na::DMatrix<i32>) {
    let mut map = HashMap::new();
    map.insert(mat, ());
}
```

Yields the following:
```
error[E0277]: the trait bound `nalgebra::base::dimension::Dynamic: std::hash::Hash` is not satisfied                                                                                                               
  --> src/lib.rs:43:9                                                                                                                                                                                              
   |                                                                                                                                                                                                               
43 |     map.insert(mat, ());                                                                                                                                                                                      
   |         ^^^^^^ the trait `std::hash::Hash` is not implemented for `nalgebra::base::dimension::Dynamic`                                                                                                        
   |                                                                                                                                                                                                               
   = note: required because of the requirements on the impl of `std::hash::Hash` for `nalgebra::base::matrix::Matrix<i32, nalgebra::base::dimension::Dynamic, nalgebra::base::dimension::Dynamic, nalgebra::base::matrix_vec::MatrixVec<i32, nalgebra::base::dimension::Dynamic, nalgebra::base::dimension::Dynamic>>`
                                                                                                                                                                                                                   
error[E0277]: the trait bound `nalgebra::base::matrix_vec::MatrixVec<i32, nalgebra::base::dimension::Dynamic, nalgebra::base::dimension::Dynamic>: std::hash::Hash` is not satisfied                               
  --> src/lib.rs:43:9                                                                                                                                                                                              
   |                                                                                                                                                                                                               
43 |     map.insert(mat, ());                                                                                                                                                                                      
   |         ^^^^^^ the trait `std::hash::Hash` is not implemented for `nalgebra::base::matrix_vec::MatrixVec<i32, nalgebra::base::dimension::Dynamic, nalgebra::base::dimension::Dynamic>`                        
   |                                                                                                                                                                                                               
   = note: required because of the requirements on the impl of `std::hash::Hash` for `nalgebra::base::matrix::Matrix<i32, nalgebra::base::dimension::Dynamic, nalgebra::base::dimension::Dynamic, nalgebra::base::matrix_vec::MatrixVec<i32, nalgebra::base::dimension::Dynamic, nalgebra::base::dimension::Dynamic>>`
```

**Comment by mwillsey** [comment]
I think it would make a great addition!

I’m using it for pattern matching submatrices in a larger matrix. So within some big matrix, count the occurrences of NxN submatrices (where N was a runtime parameter, hence the Dynamic).
So I really want to put a DMatrixSlice in the hashmap to save memory, but I think the problem is the same with regular DMatrix.

I’m not sure how common this particular operation is. I’m using it in messing around with the Wave Function Collapse algorithm. But I imagine other cases of wanting to hash small matrices.

**Comment by sebcrozet** [comment]
This is fixed by #510
